field/src/extension/ext_powers_reducer.rs

```diff
@@ -0,0 +1,79 @@
+use alloc::vec;
+use alloc::vec::Vec;
+
+use itertools::{izip, Itertools};
+
+use crate::{AbstractField, ExtensionField, Field, PackedValue};
+
+/// For an extension reducing factor a, and base vector x, can calculate sum(a^i * x_i) quickly.
+/// You must call prepare_for_width first to populate powers of a, then it is efficient to use
+/// with several base vectors. Subsequently preparing for greater widths will reuse precomputation.
+pub struct ExtensionPowersReducer<F: Field, EF> {
+    pub factor: EF,
+    next: EF,
+
+    pub powers: Vec<EF>,
+    // Say EF::D = 2 and F::Packing::WIDTH = 3
+    //    powers       vertically_packed_powers
+    // [               [
+    //   EF( 0, 1), \
+    //   EF( 2, 3),  >   [P(0,2,4), P(1,3,5)],
+    //   EF( 4, 5), /
+    //   EF( 6, 7), \
+    //   EF( 8, 9),  >   [P(6,8,10), P(7,9,11)],
+    //   EF(10,11), /
+    // ]               ]
+    vertically_packed_powers: Vec<Vec<F::Packing>>,
+}
+
+impl<F: Field, EF: ExtensionField<F>> ExtensionPowersReducer<F, EF> {
+    pub fn new(factor: EF) -> Self {
+        Self {
+            factor,
+            next: EF::one(),
+            powers: vec![],
+            vertically_packed_powers: vec![],
+        }
+    }
+
+    pub fn prepare_for_width(&mut self, width: usize) {
+        while self.powers.len() < width {
+            let old_width = self.powers.len();
+            for _ in 0..F::Packing::WIDTH {
```

Comment:
It seems that we are running this `for` loop every time the `prepare_for_width` is called. Couldn't we pre-compute this once, possibly at creation of the instance ? Moreover, `self.next` doesn't seem to be further updated..

Reply:
Powers get pushed in the next loop, in multiples of packing width